### PR TITLE
Add sound feedback to most tab control implementations from keyboard interactions

### LINKED ISSUE #18551: Ctrl-tab navigation of tabs doesn't play samples

### LINKED ISSUE #27400: No audatory feedback when switching editor modes using keybinds
### Type

Game behaviour

### Bug description

When in editor, players can switch between modes either by clicking the buttons on the top right corner of the screen or using keybinds (<kbd>F1</kbd>, <kbd>F2</kbd>, <kbd>F3</kbd>, and <kbd>F4</kbd>). Here, audatory feedback is implemented for the buttons, but not for keybinds. This is inconsistent between other parts of the game where the same audio feedback is played regardless of whether they are using mouse or keybinds (in song select, for example).

### Screenshots or videos

[untitled.webm](https://github.com/ppy/osu/assets/40858122/dcf9de1e-f97a-4555-b08c-aa04eea13bff)

### Version

2024.221.0-lazer

### Logs

[compressed-logs.zip](https://github.com/ppy/osu/files/14418404/compressed-logs.zip)

### PR DESCRIPTION
**Pull request by frenzibyte**
- Closes #18551 
- Closes #27400
- Supersedes #27448 
- [x] Depends on https://github.com/ppy/osu-framework/pull/6218
- [x] Depends on framework package

Removes usage of `HoverClickSound` and replaces it with `HoverSounds` on all tab item implementations, and overrides the newly added event `OnUserTabSelectionChanged` to play sound instead, so that audio feedback gets extended to keyboard interactions such as Ctrl-Tab or by the `SelectItem` method (used in this PR to resolve #27400). 

I will say I'm not 100% sure about the direction of this PR after reading the final diff, the fact that the logic for audio feedback is now split across `TabControl` (click) and `TabItem` (hover) feels really bad from a code quality perspective, but I don't really have any brighter idea right now.

The weirdest part of this diff would be `HoverClickSounds.PlayAsClick`, but it's necessary to share the click playback logic between `HoverClickSounds` and `TabControl`. I can remove it and let `TabControl` play the sample normally without any frequency adjustment, if that's better.

### CODE REVIEW

**Review by Joehuu — COMMENTED** [review]
Apart from playlist/multiplayer currently crashing due to no `AddEnumEntriesAutomatically` override, there is no more click sound when pressing the already selected item. Just pointing that out and maybe change to the disabled sound if its worth fixing. Otherwise changes look good to me.

**Comment by frenzibyte** [comment]
@peppy thoughts? Should a sound still be played when clicking on an already selected tab, or not play anything. I've been going with the latter until I had realised master does the former.

Also if a sound is to be played, should it be the normal click sound, or the disabled version?

**Comment by peppy** [comment]
If it's a select sound it should only play when selection changes.

**Comment by frenzibyte** [comment]
It is, there's no need for extra changes then. I'll resolve the multiplayer/playlists issues when framework is bumped here.

**Comment by frenzibyte** [comment]
I've fixed tests and resolved a few bugs with the implementation after checking previous diff. Should be in a good state for review.

**Comment by peppy** [comment]
I have real questions about some of the fixes you made...

ie. why did https://github.com/ppy/osu/pull/27657/commits/57c12a220c291670d584f955cc1e60a82bcb98c0 stop working in the first place?

why are https://github.com/ppy/osu/pull/27657/commits/7ca7328e3696645bb4f81feced976fc5def53411 / https://github.com/ppy/osu/pull/27657/commits/a603a7ecb0249eac1c46b8aa981620cf280f805e in separate commits with separate messaging? also this handling feels pretty awkward..

Given the fixes are so wide-reaching, I'm not 100% sure how to review this other than click around the UI and check if things make sounds. Is that the best direction of review?

**Comment by frenzibyte** [comment]
Main points to bring up:
 - I am moving as much `TabControl` implementations as possible to implement `OsuTabControl` for proper sound feedback on tab selection change (except for `RulesetSelector` because it plays its own sounds). I had moved the two classes in https://github.com/ppy/osu/commit/57c12a220c291670d584f955cc1e60a82bcb98c0 to inherit `OsuTabControl` but mistakingly did not update the implementations to work with [this logic in `OsuTabControl`](https://github.com/ppy/osu/blob/master/osu.Game/Graphics/UserInterface/OsuTabControl.cs#L72-L76), so I fixed that in 57c12a220c291670d584f955cc1e60a82bcb98c0.
 - With the new behaviour established in `OsuTabControl`s, using `SelectTab` has sound feedback, so in https://github.com/ppy/osu/commit/a603a7ecb0249eac1c46b8aa981620cf280f805e, I had changed that to use `Current.Value` because we don't want sound to be played in that particular scenario (executed when opening a beatmap info overlay and/or closing it).

https://github.com/ppy/osu/commit/7ca7328e3696645bb4f81feced976fc5def53411 is, as described, work around for NRT weirdness due to `TabControl` having NRT disabled while I have the affected classes (`ProfileRulesetSelector` and `BeatmapRulesetSelctor`) enabled. I can disable it back if preferred. https://github.com/ppy/osu/commit/a603a7ecb0249eac1c46b8aa981620cf280f805e also does the same (in addition to switching `SelectTab` with `Current.Value`), poor split of commits from my end.

File: osu.Game/Overlays/BeatmapListing/BeatmapSearchMultipleSelectionFilterRow.cs [hunk 1]
```diff
@@ -118,6 +119,8 @@ public MultipleSelectionFilterTabItem(T value)
                 });
             }
 
+            protected override HoverSounds CreateHoverSounds() => new HoverClickSounds(HoverSampleSet.TabSelect);
```
Comment (peppy) [inline osu.Game/Overlays/BeatmapListing/BeatmapSearchMultipleSelectionFilterRow.cs:122]: I don't know how to explain why this is needed, but it **needs an inline comment**, or it needs to be removed.

File: osu.Game/Overlays/OverlayPanelDisplayStyleControl.cs [hunk 2]
```diff
@@ -78,7 +78,7 @@ public PanelDisplayTabItem(OverlayPanelDisplayStyle value)
                         RelativeSizeAxes = Axes.Both,
                         FillMode = FillMode.Fit
                     },
-                    new HoverClickSounds()
+                    new HoverSounds()
```
Comment (peppy) [inline osu.Game/Overlays/OverlayPanelDisplayStyleControl.cs:81]: Why isn't the `TabSelect` style? And why can't it be shared in `OsuTabControl`.

**Comment by peppy** [comment]
I really hate to go through this many rounds of review for a PR like this. I'd hope it's something we could merge in 5 minutes because it's just putting sound effects in places. The reason it's hard is because we're trying to be smart and generic and reuse classes.

I wouldn't be opposed to just putting *every* instance of sound effects locally and calling it a day if that's seen as simpler. At least we wouldn't have to dance around with changes like this and waste literally hours of review time on it.

File: osu.Game/Overlays/OverlayPanelDisplayStyleControl.cs (hunk 2 above)
Comment (frenzibyte) [reply osu.Game/Overlays/OverlayPanelDisplayStyleControl.cs:81]: That's how it was previously.

File: osu.Game/Overlays/BeatmapListing/BeatmapSearchMultipleSelectionFilterRow.cs (hunk 1 above)
Comment (frenzibyte) [reply osu.Game/Overlays/BeatmapListing/BeatmapSearchMultipleSelectionFilterRow.cs:122]: Added inline comment in the alternative branch.

File: osu.Game/Overlays/BeatmapListing/FilterTabItem.cs [hunk 3]
```diff
@@ -40,12 +40,14 @@ private void load()
...
                     Text = LabelFor(Value)
                 },
-                new HoverClickSounds(HoverSampleSet.TabSelect)
+                CreateHoverSounds()
             });
 
             Enabled.Value = true;
         }
 
+        protected virtual HoverSounds CreateHoverSounds() => new HoverSounds(HoverSampleSet.TabSelect);
```
[the remark this answers is not in the document]
Comment (frenzibyte) [reply osu.Game/Overlays/BeatmapListing/FilterTabItem.cs:49]: See above.

File: osu.Game/Screens/OnlinePlay/Match/Components/MatchTypePicker.cs [hunk 4]
```diff
@@ -78,7 +80,7 @@ public GameTypePickerItem(MatchType value)
                             },
                         },
                     },
-                    new HoverClickSounds(),
+                    new HoverSounds(),
```
[the remark this answers is not in the document]
Comment (frenzibyte) [reply osu.Game/Screens/OnlinePlay/Match/Components/MatchTypePicker.cs:83]: Went the opposite way around and brought back select sounds in my alternative branch, less loc.

File: osu.Game/Screens/OnlinePlay/Match/Components/RoomAvailabilityPicker.cs [hunk 5]
```diff
@@ -75,6 +72,7 @@ public RoomAvailabilityPickerItem(RoomAvailability value)
                         Font = OsuFont.GetFont(weight: FontWeight.Bold),
                         Text = value.GetDescription(),
                     },
+                    new HoverSounds(),
```
[the remark this answers is not in the document]
Comment (frenzibyte) [reply osu.Game/Screens/OnlinePlay/Match/Components/RoomAvailabilityPicker.cs:75]: See above.

**Comment by frenzibyte** [comment]
The original intent of this PR was to move the triggering of the "tab select" sound completely from tab items and into `TabControl` itself to make it triggered by the `OnUserTabSelectChange` event (i.e. via click or via `SelectTab`). The downsides of that approach as it appeared in this PR are:
 1. Tab items still need to play hover sound, therefore `HoverSounds` still need to be implemented inside each `TabItem` implementation, making code readability quite weird due to sound not being handled in one place (i.e. https://github.com/ppy/osu/pull/27657#discussion_r1546048438)
 2. Every tab control implementation that we want to play select sound must inherit `OsuTabControl` and make sure their tab items use `HoverSounds` into `HoverClickSounds`, as to not cause double select sample playback.
 3. There are tab controls wherein their tab items have custom behaviour when clicking on them (e.g. beatmap listing "sort by" tab control).

If we go back all the way to https://github.com/ppy/osu/pull/27448, my intention there was to define an explicit method in `OsuTabControl` to play sound when selecting a tab, similar to what you're proposing by putting every instance of sound effect locally.

If you were to ask for my opinion, I'm thinking of the following as a last attempt: 
 - https://github.com/ppy/osu-framework/compare/master...frenzibyte:osu-framework:tab-control-user-events-2
 - https://github.com/ppy/osu/compare/master...frenzibyte:osu:add-sound-to-tab-selection-change-2

This branch aims to treat the downsides above by keeping "tab select" sound playback inside each tab item, avoiding the entire idea of making every tab control implement `OsuTabControl`, allowing custom tab items to still trigger select sound or completely roll off its own way of having it trigger (e.g. `OverlaySortTabControl<T>.SortTabItem`s), and greatly simplifying diff.

Notes:
 - The existence of `HoverSelectSounds` just serves as an easy way of being able to play select sample without having to manually look it up from the sample store on every tab item implementation. If it's very bad to exist, suggestions are welcomed.

**Comment by peppy** [comment]
I don't get it

2024-04-17 11 25 13@2x

Why is clicking playing a select sound?

Why do we need

`HoverClickSounds : HoverSelectSounds : HoverSounds`

Too much inheritance. Too much nonsense.

**Comment by frenzibyte** [comment]
This part that you are highlighting can be easily get ridden off by manually fetching the select sample in each tab item class, it does not play any role in the proposal.

**Comment by peppy** [comment]
Okay let's give it a try then. It seems better than this PR at lesat.

**Pull request closed without merging**